### PR TITLE
test: Plonk verifier in Haskell test

### PR DESCRIPTION
**Pull request by huitseeker**
- Added a new `.ghci` file for setting Haskell interpreter options specifically for the package structure
- Created a new `Main.hs` application for verifying Plonk algorithm proofs, utilizing the user's `HOME` environment variable and interfacing with `verify_plonk_bn254` function.
- Modified the project structure by adding a new `cabal.project` and refactoring `plonk-verify.cabal`, including updating dependencies and removing `plonk_verify` from some sections.
- Set up a dedicated `test/test.cabal` file for Haskell package configuration,
- Enhanced the GitHub CI workflow with a new Haskell test runner step.
- Adjusted the crate type setting in `Cargo.toml` from `staticlib` to `cdylib` // GHCI can't import static libs.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -21,7 +21,7 @@ sha2 = "0.10.8"
 hex = "0.4.3"
 
 [lib]
-crate-type = ["staticlib"]
+crate-type = ["cdylib"]
```
Comment (wwared) [inline Cargo.toml:24]:
```suggestion
crate-type = ["staticlib", "cdylib"]
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -36,8 +36,15 @@ jobs:
       - name: Build with Cabal
         run: cabal build
 
-      # Download and untar the required file
-      - name: Download and untar PLONK params
+      # Cache PLONK params download
+      - name: Cache PLONK params
+        uses: actions/cache@v3
```
Comment (samuelburnham) [inline .github/workflows/ci.yml:41]:
```suggestion
        uses: actions/cache@v4
        id: cache-plonk-params
```

File: test/test.cabal [hunk 3]
```diff
@@ -0,0 +1,72 @@
+cabal-version:      3.12
```
Comment (wwared) [inline test/test.cabal:1]: CI is complaining about this value ("  Errors encountered when parsing cabal file ./test/test.cabal:  test/test.cabal:0:0: error:  Unsupported cabal-version 3.12. See https://github.com/haskell/cabal/issues/4899.). What about setting it to be equal to the other `.cabal` file?
```suggestion
cabal-version:      2.4
```

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -36,8 +36,15 @@ jobs:
...
+        with:
+          path: ~/.sp1/circuits/plonk_bn254/v1.0.8-testnet
+          key: plonk-params-v1.0.8-testnet
```
Comment (samuelburnham) [inline .github/workflows/ci.yml:44]: Is there a way we can make the version variable going forward, in order to cache the params beyond `v1.0.8-testnet`? One way could be fetching the version at https://github.com/argumentcomputer/lurk-hs/blob/main/src/lib.rs#L83

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -36,8 +36,15 @@ jobs:
       - name: Build with Cabal
```
Comment (samuelburnham) [inline .github/workflows/ci.yml:36]: I think switching `haskell/actions/setup@v1` to `haskell-actions/setup@v2` above is a drop-in change

**Review by wwared — APPROVED** [review]

**Review by wwared — COMMENTED** [review]
~Unfortunately to work around the library search path issues on linux, I added `LD_LIBRARY_PATH=$(pwd)/target/release` to some CI commands. A better workaround is appreciated (e.g. using `staticlib` only and no `cdylib` in linux gets rid of this error, but leads to macos symbol failures at runtime)~

Nevermind: I've changed back the default to `staticlib`, and only explicitly build the `cdylib` on macOS. I've also updated the instructions. This seems to give better default-UX on Linux where the static linking works with no issues.

If `cabal repl` or similar needs to work, then the same `cargo rustc --release --crate-type cdylib` command should work to make the REPL work.

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -36,7 +36,7 @@ jobs:
       - uses: Swatinem/rust-cache@v2
 
       - name: Build with Cargo
-      - run: |
+        run: |
```
Comment (huitseeker) [inline .github/workflows/ci.yml:39]: 🤦

File: README.md [hunk 7]
```diff
@@ -54,7 +54,11 @@ To run the example Haskell application, execute:
 cabal run test
 ```
 
-If that fails with "error while loading shared libraries", try instead: (for using `target/release/libplonk_verify.so`)
+## Troubleshooting
+
+If you get "error while loading shared libraries", try:
 ```bash
-LD_LIBRARY_PATH=$(pwd)/target/release cabal run test
+cargo clean
```
Comment (huitseeker) [inline README.md:61]: 🙏

**Review by huitseeker — COMMENTED** [review]
Thanks @wwared

**Review by samuelburnham — APPROVED** [review]

**Pull request merged**